File: backend/src/main/java/com/back/team11/domain/cafe/entity/Cafe.java
package com.back.team11.domain.cafe.entity;

import com.back.team11.domain.member.entity.Member;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Getter
@Entity
@Builder  // cafe-CRUD 추가: 필드명 기반 명시적 생성, 선택 필드 생략 가능
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class) // createdAt, updatedAt 자동 관리
@Table(name = "cafe")
public class Cafe {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    private Member member;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(nullable = false)
    private String address;

    @Column(nullable = false, precision = 10, scale = 7)
    private BigDecimal latitude;

    @Column(nullable = false, precision = 11, scale = 7)
    private BigDecimal longitude;

    @Column(length = 20)
    private String phone;

    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private CafeType type;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private Franchise franchise;

    @Column(nullable = false)
    private Boolean hasToilet;

    @Column(nullable = false)
    private Boolean hasOutlet;

    @Column(nullable = false)
    private Boolean hasWifi;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private FloorCount floorCount;

    @Column(nullable = false)
    private Boolean hasSeparateSpace;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private CongestionLevel congestionLevel;

    @Column(length = 500)
    private String imageUrl;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private CafeStatus status;

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(nullable = false)
    private LocalDateTime updatedAt;


    // ─────────────────────────────────────────────
    // 정적 팩토리 메서드 - new 대신 static 사용 (필수 값 누락을 방지), 도메인 로직을 엔티티에 캡슐화
    // member 없이 생성 - 관리자가 직접 등록하기에 필요 X
    // ─────────────────────────────────────────────
    public static Cafe createByAdmin(
            String name,
            String address,
            BigDecimal latitude,
            BigDecimal longitude,
            String phone,
            String description,
            CafeType type,
            Franchise franchise,
            Boolean hasToilet,
            Boolean hasOutlet,
            Boolean hasWifi,
            FloorCount floorCount,
            Boolean hasSeparateSpace,
            CongestionLevel congestionLevel,
            String imageUrl
    ) {
        return Cafe.builder()
                .name(name)
                .address(address)
                .latitude(latitude)
                .longitude(longitude)
                .phone(phone)
                .description(description)
                .type(type)
                .franchise(franchise)
                .hasToilet(hasToilet)
                .hasOutlet(hasOutlet)
                .hasWifi(hasWifi)
                .floorCount(floorCount)
                .hasSeparateSpace(hasSeparateSpace)
                .congestionLevel(congestionLevel)
                .imageUrl(imageUrl)
                .status(CafeStatus.APPROVED) // status는 APPROVED 고정: 관리자 직접 등록 → 즉시 승인
                .build();
    } // @Builder 사용으로 생성자 대신 정적 팩토리 메서드로 명시적 생성, 필드명 기반, 선택 필드 생략 가능


    // ─────────────────────────────────────────────
    // PATCH 수정 메서드 - null인 필드는 기존값 유지
    // member, status, createdAt, updatedAt는 수정 대상 아님 (status = APPROVED 고정)
    // 엔티티 자신의 필드를 변경하는 메서드. 변경 결과를 반환할 필요 X
    // ─────────────────────────────────────────────
    public void updateByAdmin(
            String name,
            String address,
            BigDecimal latitude,
            BigDecimal longitude,
            String phone,
            String description,
            CafeType type,
            Franchise franchise,
            Boolean hasToilet,
            Boolean hasOutlet,
            Boolean hasWifi,
            FloorCount floorCount,
            Boolean hasSeparateSpace,
            CongestionLevel congestionLevel,
            String imageUrl
    ) {
        if (name != null)              this.name = name;
        if (address != null)           this.address = address;
        if (latitude != null)          this.latitude = latitude;
        if (longitude != null)         this.longitude = longitude;
        if (phone != null)             this.phone = phone;
        if (description != null)       this.description = description;
        if (type != null)              this.type = type;
        if (franchise != null)         this.franchise = franchise;
        if (hasToilet != null)         this.hasToilet = hasToilet;
        if (hasOutlet != null)         this.hasOutlet = hasOutlet;
        if (hasWifi != null)           this.hasWifi = hasWifi;
        if (floorCount != null)        this.floorCount = floorCount;
        if (hasSeparateSpace != null)  this.hasSeparateSpace = hasSeparateSpace;
        if (congestionLevel != null)   this.congestionLevel = congestionLevel;
        if (imageUrl != null)          this.imageUrl = imageUrl;
    }


    // ─────────────────────────────────────────────
    // 정적 팩토리 메서드 - 사용자 제보
    // member 연결, status는 PENDING 고정 (관리자 심사 대기)
    // ─────────────────────────────────────────────
    public static Cafe createByUser(
            Member member,
            String name, String address,
            BigDecimal latitude, BigDecimal longitude,
            String phone, String description,
            CafeType type, Franchise franchise,
            Boolean hasToilet, Boolean hasOutlet, Boolean hasWifi,
            FloorCount floorCount, Boolean hasSeparateSpace,
            CongestionLevel congestionLevel, String imageUrl
    ) {
        return Cafe.builder()
                .member(member)
                .name(name)
                .address(address)
                .latitude(latitude)
                .longitude(longitude)
                .phone(phone)
                .description(description)
                .type(type)
                .franchise(franchise)
                .hasToilet(hasToilet)
                .hasOutlet(hasOutlet)
                .hasWifi(hasWifi)
                .floorCount(floorCount)
                .hasSeparateSpace(hasSeparateSpace)
                .congestionLevel(congestionLevel)
                .imageUrl(imageUrl)
                .status(CafeStatus.PENDING) // 사용자 제보 → 관리자 심사 대기
                .build();
    }


    // 승인 처리
    public void approve() {
        this.status = CafeStatus.APPROVED;
    }

    // 거절 처리
    public void reject() {
        this.status = CafeStatus.REJECTED;
    }

}

File: backend/src/main/java/com/back/team11/domain/cafe/entity/Cafe.kt
package com.back.team11.domain.cafe.entity

import com.back.team11.domain.member.entity.Member
import jakarta.persistence.*
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
@EntityListeners(AuditingEntityListener::class) // createdAt, updatedAt 자동 관리
@Table(name = "cafe")
class Cafe(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    var member: Member? = null,

    @Column(nullable = false, length = 100)
    var name: String,

    @Column(nullable = false)
    var address: String,

    @Column(nullable = false, precision = 10, scale = 7)
    var latitude: BigDecimal,

    @Column(nullable = false, precision = 11, scale = 7)
    var longitude: BigDecimal,

    @Column(length = 20)
    var phone: String? = null,

    var description: String? = null,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    var type: CafeType,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    var franchise: Franchise,

    @Column(nullable = false)
    var hasToilet: Boolean,

    @Column(nullable = false)
    var hasOutlet: Boolean,

    @Column(nullable = false)
    var hasWifi: Boolean,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    var floorCount: FloorCount,

    @Column(nullable = false)
    var hasSeparateSpace: Boolean,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    var congestionLevel: CongestionLevel,

    @Column(length = 500)
    var imageUrl: String? = null,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    var status: CafeStatus,

    @CreatedDate
    @Column(nullable = false, updatable = false)
    var createdAt: LocalDateTime? = null,

    @LastModifiedDate
    @Column(nullable = false)
    var updatedAt: LocalDateTime? = null,
) {
    companion object {
        // ─────────────────────────────────────────────
        // 정적 팩토리 메서드 - 관리자 직접 등록
        // member 없이 생성, status는 APPROVED 고정
        // ─────────────────────────────────────────────
        fun createByAdmin(
            name: String,
            address: String,
            latitude: BigDecimal,
            longitude: BigDecimal,
            phone: String?,
            description: String?,
            type: CafeType,
            franchise: Franchise,
            hasToilet: Boolean,
            hasOutlet: Boolean,
            hasWifi: Boolean,
            floorCount: FloorCount,
            hasSeparateSpace: Boolean,
            congestionLevel: CongestionLevel,
            imageUrl: String?,
        ): Cafe = Cafe(
            name = name,
            address = address,
            latitude = latitude,
            longitude = longitude,
            phone = phone,
            description = description,
            type = type,
            franchise = franchise,
            hasToilet = hasToilet,
            hasOutlet = hasOutlet,
            hasWifi = hasWifi,
            floorCount = floorCount,
            hasSeparateSpace = hasSeparateSpace,
            congestionLevel = congestionLevel,
            imageUrl = imageUrl,
            status = CafeStatus.APPROVED, // 관리자 직접 등록 → 즉시 승인
        )

        // ─────────────────────────────────────────────
        // 정적 팩토리 메서드 - 사용자 제보
        // member 연결, status는 PENDING 고정 (관리자 심사 대기)
        // ─────────────────────────────────────────────
        fun createByUser(
            member: Member,
            name: String,
            address: String,
            latitude: BigDecimal,
            longitude: BigDecimal,
            phone: String?,
            description: String?,
            type: CafeType,
            franchise: Franchise,
            hasToilet: Boolean,
            hasOutlet: Boolean,
            hasWifi: Boolean,
            floorCount: FloorCount,
            hasSeparateSpace: Boolean,
            congestionLevel: CongestionLevel,
            imageUrl: String?,
        ): Cafe = Cafe(
            member = member,
            name = name,
            address = address,
            latitude = latitude,
            longitude = longitude,
            phone = phone,
            description = description,
            type = type,
            franchise = franchise,
            hasToilet = hasToilet,
            hasOutlet = hasOutlet,
            hasWifi = hasWifi,
            floorCount = floorCount,
            hasSeparateSpace = hasSeparateSpace,
            congestionLevel = congestionLevel,
            imageUrl = imageUrl,
            status = CafeStatus.PENDING, // 사용자 제보 → 관리자 심사 대기
        )
    }

    // ─────────────────────────────────────────────
    // PATCH 수정 메서드 - null인 필드는 기존값 유지
    // member, status, createdAt, updatedAt는 수정 대상 아님
    // ─────────────────────────────────────────────
    fun updateByAdmin(
        name: String?,
        address: String?,
        latitude: BigDecimal?,
        longitude: BigDecimal?,
        phone: String?,
        description: String?,
        type: CafeType?,
        franchise: Franchise?,
        hasToilet: Boolean?,
        hasOutlet: Boolean?,
        hasWifi: Boolean?,
        floorCount: FloorCount?,
        hasSeparateSpace: Boolean?,
        congestionLevel: CongestionLevel?,
        imageUrl: String?,
    ) {
        name?.let { this.name = it }
        address?.let { this.address = it }
        latitude?.let { this.latitude = it }
        longitude?.let { this.longitude = it }
        phone?.let { this.phone = it }
        description?.let { this.description = it }
        type?.let { this.type = it }
        franchise?.let { this.franchise = it }
        hasToilet?.let { this.hasToilet = it }
        hasOutlet?.let { this.hasOutlet = it }
        hasWifi?.let { this.hasWifi = it }
        floorCount?.let { this.floorCount = it }
        hasSeparateSpace?.let { this.hasSeparateSpace = it }
        congestionLevel?.let { this.congestionLevel = it }
        imageUrl?.let { this.imageUrl = it }
    }

    // 승인 처리
    fun approve() { status = CafeStatus.APPROVED }

    // 거절 처리
    fun reject() { status = CafeStatus.REJECTED }
}